backend/core/src/application-flagged-sets/application-flagged-sets-cronjob-consumer.ts

```diff
@@ -0,0 +1,293 @@
+
+    let afses = await this.afsRepository
+      .createQueryBuilder("afs")
+      .leftJoin("afs.applications", "applications")
+      .select(["afs", "applications.id"])
+      .where(`afs.listing_id = :listingId`, { listingId: application.listingId })
+      .getMany()
+
+    afses = afses.filter((afs) => afs.applications.map((app) => app.id).includes(application.id))
+
+    const afsesToBeSaved: Array<ApplicationFlaggedSet> = []
+    const afsesToBeRemoved: Array<ApplicationFlaggedSet> = []
+
+    for (const afs of afses) {
+      afs.status = FlaggedSetStatus.flagged
+      afs.resolvedTime = null
+      afs.resolvingUser = null
+
+      const applicationIndex = afs.applications.findIndex((app) => app.id === application.id)
+
+      if (applicationIndex == -1) {
+        continue
+      }
+
+      afs.applications.splice(applicationIndex, 1)
+
+      if (afs.applications.length > 1) {
+        afsesToBeSaved.push(afs)
+      } else {
+        afsesToBeRemoved.push(afs)
+      }
+    }
+
+    await this.afsRepository.save(afsesToBeSaved)
+    await this.afsRepository.remove(afsesToBeRemoved)
+
+    await this.addApplication(application, rule)
+  }
+
+  async addApplication(newApplication: Application, rule: Rule) {
+    const applicationsMatchingRule = await this.fetchDuplicatesMatchingRule(newApplication, rule)
+
+    if (applicationsMatchingRule.length == 0) {
+      return
+    }
+
+    const afses = await this.afsRepository
+      .createQueryBuilder("afs")
+      .leftJoin("afs.applications", "applications")
+      .select(["afs", "applications.id"])
+      .where(`afs.ruleKey = :ruleKey`, { ruleKey: this.getRuleKeyForRule(newApplication, rule) })
+      .getMany()
+
+    if (afses.length == 0) {
+      await this.afsRepository.save({
+        rule: rule,
+        ruleKey: this.getRuleKeyForRule(newApplication, rule),
+        resolvedTime: null,
+        resolvingUser: null,
+        status: FlaggedSetStatus.flagged,
+        applications: [newApplication, ...applicationsMatchingRule],
+        listing: { id: newApplication.listingId },
+      })
+    } else if (afses.length == 1) {
+      const afs = afses[0]
+
+      if (!afs.applications.map((app) => app.id).includes(newApplication.id)) {
+        afs.applications.push(newApplication)
+        await this.afsRepository.save(afs)
+      }
+    } else {
+      console.error(
+        "There should be up to one AFS matching a rule for given application, " +
+          "probably a logic error when creating AFSes"
+      )
+    }
+  }
+
+  private async fetchDuplicatesMatchingRule(application: Application, rule: Rule) {
+    switch (rule) {
+      case Rule.nameAndDOB:
+        return await this.fetchDuplicatesMatchingNameAndDOBRule(application)
+      case Rule.email:
+        return await this.fetchDuplicatesMatchingEmailRule(application)
+    }
+  }
+
+  private async fetchDuplicatesMatchingEmailRule(newApplication: Application) {
+    return await this.applicationRepository.find({
```

Comment:
something pretty similar to this using the afsrepository was what was causing our backend to crash under larger application loads, so we may want to convert this into a querybuilder that we construct ourselves

Reply:
Noting here for #2959